File: adv-langs/js-adv/precedence.js
const findword = (arr) => {
    let subs = arr.map(el => el.split('>'))
    let res = []
    let count = 20
    
    while (subs.length & count){
        if (!res.length){
            res.push(subs[0])
            subs.shift()
        }
        for (pair of subs) {
            count--;
            if (res[0][0] === pair[1]) {
                res.unshift(pair)
                subs = subs.filter(el => el !== pair)
            }
            else if (res.slice(-1)[0][1] === pair[0]){
                res.push(pair)
                subs = subs.filter(el => el !== pair)
            }          
        }
    }
    const p1 = (res.slice(0, -1).map(el => el[0])).join('')
    const p2 = res.slice(-1)[0].join('')
    return p1 + p2
}

r = findword(["I>N", "A>I", "P>A", "S>P"]) 
// # sp pa ai in
console.log(r)

r2 = findword(["R>T", "A>L", "P>O", "O>R", "G>A", "T>U", "U>G"])
console.log(r2)

File: adv-langs/py-adv/2025_may_week_2/precedence.py
def findWord(arr):
    # split arr into pairs
    subs = [e.split('>') for e in arr]
    
    result = []
    count = 20

    # loop through 
    while subs and count != 0:
        count -= 1
        if not result:
            result.append(subs[0])
            subs.remove(subs[0])
        else:
            for pair in subs:
                # 1st char in result starting pair equal to 2nd char in pair
                if result[0][0] == pair[1]:
                    # inser at beggining and remove from subs
                    result.insert(0, pair)
                    subs.remove(pair)
                # final char of result equals 1st char on pair 
                elif result[-1][1] == pair[0]:
                    # insert at end and remove from subs
                    result.append(pair)
                    subs.remove(pair)
    
    r = [el[0] for el in result[:-1]] + result[-1]

    return ''.join(r)

r = findWord(["I>N", "A>I", "P>A", "S>P"]) 
# sp pa ai in
print(r)

r2 = findWord(["R>T", "A>L", "P>O", "O>R", "G>A", "T>U", "U>G"])
print(r2)
# po or rt tu ug ga al
# findWord(["P>E", "E>R", "R>U"])

# findWord(["U>N", "G>A", "R>Y", "H>U", "N>G", "A>R"]) // HUNGARY
# findWord(["I>F", "W>I", "S>W", "F>T"]) // SWIFT
r3 = findWord(["W>I", "R>L", "T>Z", "Z>E", "S>W", "E>R", "L>A", "A>N", "N>D", "I>T"])
print(r3)